crates/ide/src/hover/render.rs

```diff
@@ -500,3 +498,94 @@ fn local(db: &RootDatabase, it: hir::Local) -> Option<Markup> {
     };
     markup(None, desc, None)
 }
+
+struct KeywordHint {
+    description: String,
+    documentation: String,
+    actions: Vec<HoverAction>,
+}
+
+impl KeywordHint {
+    fn new(description: String, documentation: String) -> Self {
+        Self { description, documentation, actions: Vec::default() }
+    }
+}
+
+/// Panics
+/// ------
+/// `token` is assumed to:
+/// - have a parent, and
+/// - be a keyword
+fn keyword_hints<'t>(sema: &Semantics<RootDatabase>, token: &'t SyntaxToken) -> KeywordHint {
+    let parent = token.parent().expect("token was assumed to have a parent, but had none");
+
+    macro_rules! create_hint {
+        ($ty_info:expr, $doc:expr) => {{
+            let documentation = $doc;
+            match $ty_info {
+                Some(ty) => {
+                    let mut targets: Vec<hir::ModuleDef> = Vec::new();
+                    let mut push_new_def = |item: hir::ModuleDef| {
+                        if !targets.contains(&item) {
+                            targets.push(item);
+                        }
+                    };
+                    walk_and_push_ty(sema.db, &ty.original, &mut push_new_def);
+
+                    let ty = ty.adjusted();
+                    let description = format!("{}: {}", token.text(), ty.display(sema.db));
+
+                    KeywordHint {
+                        description,
+                        documentation,
+                        actions: vec![HoverAction::goto_type_from_targets(sema.db, targets)],
+                    }
+                }
+                None => KeywordHint {
+                    description: token.text().to_string(),
+                    documentation,
+                    actions: Vec::new(),
+                },
+            }
+        }};
+    }
```

Comment:
This does not look like it needs to be a macro but a function or (even better in my eyes) closure instead.